javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,373 @@
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(nextNode: Node[E], e: E): Unit = {
+    if (nextNode eq head) addFirst(e)
+    else if (nextNode eq null) addLast(e)
+    else {
+
+      val node = new Node(e, prev = nextNode.prev, next = nextNode)
+      nextNode.prev.next = node
+      nextNode.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) removeFirst()
+    else if (node eq last) removeLast()
+    else {
+
+      node.prev.next = node.next
+      node.next.prev = node.prev
+
+      _size -= 1
+
+      node.value
+    }
+  }
+
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    removeNode(getNodeAt(index))
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    removeFirst()
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E =
+    head.value
+
+  def peekLast(): E =
+    peekFirst()
+
+  def pollFirst(): E =
+    removeFirst()
+
+  def pollLast(): E =
+    removeLast()
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  private def _removeOccurrence(iter: Iterator[E], o: Any): Boolean = {
+    var changed = false
+    val e = o.asInstanceOf[E]
+    while (iter.hasNext() && !changed) {
+      if (iter.next() === e) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean =
+    _removeOccurrence(iterator(), o)
+
+  def removeLastOccurrence(o: Any): Boolean =
+    _removeOccurrence(descendingIterator(), o)
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new ListIterator[E] {
+
+      private var last = -1
+      private var i = index
+
+      private var currentNode: Node[E] =
+        if (index == size) null
+        else getNodeAt(index)
+
+      private var lastNode: Node[E] =
+        if (currentNode eq null) getNodeAt(size - 1)
+        else null
+
+      def hasNext(): Boolean =
+        i < size
+
+      def next(): E = {
+        if (i >= size) throw new NoSuchElementException()
+        last = i
+        i += 1
+
+        if (currentNode eq null)
+            currentNode = lastNode
+
+        lastNode = currentNode
+        currentNode = currentNode.next
+
+        lastNode.value
+      }
+
+      def hasPrevious(): Boolean =
+        i > 0
+
+      def previous(): E = {
+        if (i < 0) throw new NoSuchElementException()
+        i -= 1
+        last = i
+
+        if (currentNode eq null)
+          currentNode = lastNode
```

Comment:
Again, assuming I understood correctly `currentNode`, then this is equivalent to
```scala
currentNode = LinkedList.this.last
```
which does not depend on the value of `lastNode`, thereby reducing the cognitive load about this field which is hard to grasp.